### PR TITLE
feat: support columar encoding for datums

### PR DESCRIPTION
**Pull request by ShiKaiWi**
## Rationale
The encoding format for wal log is not compact, which imposes too much overhead on the underlying wal service.

## Detailed Changes
- Introduce the columnar encoding framework sub-module in the codec module.
- Provide some basic examples of implementation for ValuesEncoder/ValuesDecoder.

## Test Plan
New unit tests.

## Future works
Implement specific encoders for every data type.

### CODE REVIEW

File: components/codec/src/columnar/mod.rs [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+
+/// The implementation for [`ValuesEncoder`].
+struct ValuesEncoderImpl;
+
+/// The implementation for [`ValuesDecoder`].
+struct ValuesDecoderImpl;
+
+#[derive(Clone, Debug)]
+pub struct ColumnarEncoder {
+    column_idx: u32,
```
Comment (jiacai2050) [inline components/codec/src/columnar/mod.rs:105]: ColumnId is better than idx.

File: components/codec/src/columnar/mod.rs [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+    }
+
+    fn compute_num_datums<'a, I>(&mut self, datums: &I) -> usize
+    where
+        I: Iterator<Item = DatumView<'a>> + Clone,
+    {
+        if let Some(v) = self.num_datums {
+            v
+        } else {
+            let num_datums = datums.clone().count();
```
Comment (jiacai2050) [inline components/codec/src/columnar/mod.rs:136]: Invoke Iterator's size_hint?

File: components/codec/src/columnar/mod.rs [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+    /// Refer to the [encode](ColumnarEncoder::encode) method.
+    #[inline]
+    const fn header_size() -> usize {
+        1 + 1 + 4 + 4 + 4
+    }
+
+    /// The layout of the final serialized bytes:
+    /// ```plaintext
+    /// +-------------+----------------+-----------------+-----------------+----------------+---------------+---------------------+
+    /// | version(u8) | datum_kind(u8) | column_idx(u32) | num_datums(u32) | num_nulls(u32) | nulls_bit_set | non-null data block |
```
Comment (jiacai2050) [inline components/codec/src/columnar/mod.rs:162]: How about put nulls_bit_set at the end of buf, then when decode, we  will combine decode_without_nulls and decode_with_nulls into one function.

File: components/codec/src/columnar/mod.rs (hunk 2 above)
Comment (ShiKaiWi) [reply components/codec/src/columnar/mod.rs:136]: What we need is an exact count of the iterator, and I guess `size_hint` can't meet this requirement.

File: components/codec/src/columnar/mod.rs (hunk 1 above)
Comment (ShiKaiWi) [reply components/codec/src/columnar/mod.rs:105]: 👍

File: components/codec/src/columnar/mod.rs (hunk 3 above)
Comment (ShiKaiWi) [reply components/codec/src/columnar/mod.rs:162]: I guess it's good to separate the two methods according to whether null values are contained.

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Pull request merged**